core/src/avm2/array.rs

```diff
@@ -11,8 +15,197 @@ use std::{cmp::max, ops::RangeBounds};
 /// the prototype.
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
-pub struct ArrayStorage<'gc> {
-    storage: Vec<Option<Value<'gc>>>,
+pub enum ArrayStorage<'gc> {
+    Dense(Vec<Option<Value<'gc>>>, usize),
+    Sparse(BTreeMap<usize, Value<'gc>>, usize),
```

Comment:
Likewise the `usize` for `Sparse` should also be documented - it serves as an explicit ECMAScript Array.length, because we can't recover that information from what's in the `BTreeMap`. This notably is *different* from the `Dense` length parameter's purpose.